exercises/dominoes/src/test/java/DominoesTest.java

```diff
@@ -0,0 +1,201 @@
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+	}
+	
+	@Ignore("Remove to run test")
+	@Test
+	public void disconnectedDoubleLoopTest() {
+		Dominoes dominoes = new Dominoes();
+		
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(2, 1), new Domino(3, 4), new Domino(4, 3)};
+		ArrayList<Domino> dominoesArray = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		expectedException.expect(IllegalArgumentException.class);
+		expectedException.expectMessage("No domino chain found.");
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+	}
+	
+	@Ignore("Remove to run test")
+	@Test
+	public void disconnectedSingleIsolatedTest() {
+		Dominoes dominoes = new Dominoes();
+		
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(2, 3), new Domino(3, 1), new Domino(4, 4)};
+		ArrayList<Domino> dominoesArray = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		expectedException.expect(IllegalArgumentException.class);
+		expectedException.expectMessage("No domino chain found.");
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+	}
+	
+	@Ignore("Remove to run test")
+	@Test
+	public void needBacktrackTest() {
+		Dominoes dominoes = new Dominoes();
+		
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(2, 3), new Domino(3, 1), new Domino(2, 4), new Domino(4, 2)};
+		ArrayList<Domino> dominoesArray = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+		
+		assertValidChain(dominoesArray, chain);
+	}
+	
+	@Ignore("Remove to run test")
+	@Test
+	public void separateLoopsTest() {
+		Dominoes dominoes = new Dominoes();
+		
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(2, 3), new Domino(3, 1), new Domino(1, 1), new Domino(2, 2), new Domino(3, 3)};
+		ArrayList<Domino> dominoesArray = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+		
+		assertValidChain(dominoesArray, chain);
+	}
+	
+	@Ignore("Remove to run test")
+	@Test
+	public void nineElementsTest() {
+		Dominoes dominoes = new Dominoes();
+		Domino[] dominoesList = {new Domino(1, 2), new Domino(5, 3), new Domino(3, 1), new Domino(1, 2), new Domino(2, 4), new Domino(1, 6), new Domino(2, 3), new Domino(3, 4), new Domino(5, 6)};
+		ArrayList<Domino> dominoesArray = new ArrayList<>(Arrays.asList(dominoesList));
+		
+		ArrayList<Domino> chain = dominoes.formChain(dominoesArray);
+		
+		assertValidChain(dominoesArray, chain);
+	}
+	
+	private void assertValidChain(ArrayList<Domino> inputDominoes, ArrayList<Domino> outputDominoes) {
+		assertSameDominoes(inputDominoes, outputDominoes);
+		assertEndDominoesMatch(outputDominoes);
+		assertConsecutiveDominoes(outputDominoes);
+	}
+	
+	private void assertEndDominoesMatch(ArrayList<Domino> outputDominoes) {
+		assertEquals("The end dominoes are not the same.", outputDominoes.get(0).getLeft(), outputDominoes.get(outputDominoes.size() - 1).getRight());
```

Comment:
Maybe something like:
```
int leftValueOfFirstDomino = outputDominoes.get(0).getLeft();
int rightValueOfLastDomino = outputDominoes.get(outputDominoes.size() -1).getRight();
String errorMessage = "The left value of the first domino ("
    + leftValueOfFirstDomino
    + ") needs to match the right value of the last domino ("
    + rightValueOfLastDomino
    + ")."
```
That way the user gets told exactly which values are wrong and why 🙂 

When a test fails the user will want to know:
- What is their code returning and what should it be returning -> the normal assert error message will tell them this
- Why is the current value wrong -> your current error message will tell them this
- Which variables are wrong -> this is what's currently missing. For example in this case they'll know that the end dominoes are not the same and they'll know the values of both ends, but not which value is which end. They could figure it out by inspecting the test code and knowing how `assertEquals` formats its error messages, but they shouldn't need to do that 🙂